### DISCUSSION TITLE
Using file system to save device settings...

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#17712] [General]
In my driver, I implemented the code from xdrv_122 (98) to save my settings and it appears to work just fine...
But I did notice that if I look at the file in "Manage File System" on the device, and look at the file date, its:
~~~
1970-01-01t00:00:01
~~~
I also do not yet understand the need for this code, what its trying to tell me, in the function below:
~~~
if (Settings->version < 0x01010100) {
      AddLog(LOG_LEVEL_INFO, PSTR("DRV: Update oldest version restore"));

    }
    if (Settings->version < 0x01010101) {
      AddLog(LOG_LEVEL_INFO, PSTR("DRV: Update old version restore"));

    }
~~~
~~~
void DrvDemoSettingsDelta(void) {
  // Fix possible setting deltas

  if (Drv122Settings.version != DRV122_VERSION) {      // Fix version dependent changes

    if (Settings->version < 0x01010100) {
      AddLog(LOG_LEVEL_INFO, PSTR("DRV: Update oldest version restore"));

    }
    if (Settings->version < 0x01010101) {
      AddLog(LOG_LEVEL_INFO, PSTR("DRV: Update old version restore"));

    }

    // Set current version and save settings
    Drv122Settings.version = DRV122_VERSION;
    DrvDemoSettingsSave();
  }
}
~~~

**Comment by arendst** [comment]
Not sure why your date isn't set but if at time of writing your clock isn't in (NTP) sync it's likely to fail.

I use the same code in drivers ``xdrv_86_esp32_sonoff_spm.ino`` and ``xdrv_87_esp32_sonoff_tm1621.ino``. This may give you more hints about the code question you're asking.

Once your driver is released you most probably want to extent it some day. Without version management on the settings you''ll encounter issues with users running the old version and your new version; you want to add settings without overwriting changed user settings. Then you add your settings as delta settings instead of resetting current user settings. again take a look at the above files.

Looking at the end of file ``settings.ino`` explains why there are so many migration steps when upgrading from an old version to the latest version; only the intermediate steps fix the deltas in setting changes as later versions will drop old deltas in favour of freeing code space.
